### PR TITLE
log: add histogram metrics for gradients

### PR DESCRIPTION
**Pull request by hasan-yaman**
Inspired by [wandb.watch](https://docs.wandb.ai/ref/python/watch/).
Add support for exporting histogram metrics.
Export gradients to see vanishing / gradient gradients.

### CODE REVIEW

**Review by takuseno — COMMENTED** [review]
@hasan-yaman Thank you for your another PR! This feature is really nice. Please let me share my thoughts here:
- Having two different methods for the same purpose is a little redundant.
- `write_histogram` is currently only used to monitor gradients.

Here is my proposal. We can remove `write_histogram` and just keep `watch_model` alone. `watch_model` is called at every `update` steps. For wandb, it calls `self.run.watch` for the first time, but does nothing after that. For `FileAdapter` and `TensorboardAdapter`, it computes gradient histograms there every `gradient_loggin_steps`. What do you think?

**Comment by hasan-yaman** [comment]
@takuseno Thanks for the comments! 
This way code looks simpler.

File: d3rlpy/logging/file_adapter.py [hunk 1]
```diff
@@ -76,6 +82,25 @@ def close(self) -> None:
...
+    def watch_model(
+        self,
+        epoch: int,
+        step: int,
+        logging_steps: Optional[int],
+        algo: TorchModuleProtocol,
+    ) -> None:
+        if logging_steps is not None and step % logging_steps == 0:
+            for name, grad in algo.impl.modules.get_gradients():
+                path = os.path.join(self._logdir, f"{name}.csv")
```
Comment (takuseno) [inline d3rlpy/logging/file_adapter.py:94]:
```
path = os.path.join(self._logdir, f"{name}_grad.csv")
```

File: d3rlpy/algos/qlearning/base.py [hunk 2]
```diff
@@ -520,6 +525,8 @@ def fitter(
         # save hyperparameters
         save_config(self, logger)
 
+        logger.watch_model(0, 0, gradient_logging_steps, self)
```
Comment (hasan-yaman) [inline d3rlpy/algos/qlearning/base.py:528]: I don't like it but it is required for wandb watch. Without this line wand doesn't track the first epoch.

File: d3rlpy/torch_utility.py [hunk 3]
```diff
@@ -388,6 +391,19 @@ def reset_optimizer_states(self) -> None:
             if isinstance(v, torch.optim.Optimizer):
                 v.state = collections.defaultdict(dict)
 
+    def get_torch_modules(self) -> List[nn.Module]:
+        torch_modules: List[nn.Module] = []
+        for v in asdict_without_copy(self).values():
```
Comment (takuseno) [inline d3rlpy/torch_utility.py:396]: It's nice to return both `key` and `value` for `get_gradients`:
```py
torch_modules = {}
for k, v in asdict_without_copy(self).values():
    if isinstance(v, nn.Module):
                torch_modules[k] = v
return torch_modules
```

File: d3rlpy/torch_utility.py [hunk 4]
```diff
@@ -388,6 +391,19 @@ def reset_optimizer_states(self) -> None:
...
+            if isinstance(v, nn.Module):
+                torch_modules.append(v)
+        return torch_modules
+
+    def get_gradients(self) -> Iterator[Tuple[str, Float32NDArray]]:
+        for module in self.get_torch_modules():
```
Comment (takuseno) [inline d3rlpy/torch_utility.py:402]: Concatenate module name and parameter names otherwise names conflict.
```py
for module_name, module in self.get_torch_modules().items():
...
        yield f"{module_name}.{name}", parameter.grad.cpu().detach().numpy()
```

**Review by takuseno — CHANGES_REQUESTED** [review]
Thanks for the change! I noticed that currently, name of parameters are conflicting without their parent module names. I left some suggestions to resolve this.

File: d3rlpy/logging/file_adapter.py (hunk 1 above)
Comment (hasan-yaman) [reply d3rlpy/logging/file_adapter.py]: [0fb02e7](https://github.com/takuseno/d3rlpy/pull/424/commits/0fb02e79c8bdeeb6a47c12a07b746f7a7b7aa47d)

File: d3rlpy/torch_utility.py (hunk 4 above)
Comment (hasan-yaman) [reply d3rlpy/torch_utility.py]: 5b8bb3c

File: d3rlpy/torch_utility.py (hunk 3 above)
Comment (hasan-yaman) [reply d3rlpy/torch_utility.py]: 5b8bb3c

File: d3rlpy/logging/wandb_adapter.py [hunk 5]
```diff
@@ -52,6 +58,21 @@ def close(self) -> None:
...
+        logging_steps: Optional[int],
+        algo: TorchModuleProtocol,
+    ) -> None:
+        if not self._is_model_watched:
+            self.run.watch(
+                tuple(algo.impl.modules.get_torch_modules().values()),
+                log="gradients",
+                log_freq=logging_steps,
+            )
+            self._is_model_watched = True
```
Comment (hasan-yaman) [inline d3rlpy/logging/wandb_adapter.py:74]: Instead of `watch`, we can use 
```python
self.run.log({"name": wandb.Histogram(...) })
```

not sure which direction I should choose

File: d3rlpy/algos/qlearning/base.py (hunk 2 above)
Comment (hasan-yaman) [reply d3rlpy/algos/qlearning/base.py:528]: Can be fixed via https://github.com/takuseno/d3rlpy/pull/424#discussion_r1798196290

**Comment by takuseno** [comment]
@hasan-yaman Thanks for the update! The implementation looks good to me. It seems that CI complains about some typing issues. Once they're fixed, let merge your PR 😄

File: d3rlpy/logging/wandb_adapter.py (hunk 5 above)
Comment (takuseno) [reply d3rlpy/logging/wandb_adapter.py:74]: For now, I'm okay with the either way. The current implementation also looks good to me.

**Comment by hasan-yaman** [comment]
@takuseno couldn't understand and fix the typing issues. am i missing something?

**Comment by takuseno** [comment]
@hasan-yaman Can you simply add `type: ignore` just to skip errors? I can fix them later once we merge this. In this PR, I'd like to suppress errors.

**Comment by hasan-yaman** [comment]
@takuseno done!

**Review by takuseno — APPROVED** [review]
LGTM. Thank you for your contribution!
